source/log_channel.c

```diff
@@ -0,0 +1,260 @@
+        /*
+         * This feels like it should be a fatal error, as it's invisible to the user and it will result
+         * in an endless backup of log lines, assuming they are able to be created.
+         */
+        return;
+    }
+
+    while(true) {
+        aws_mutex_lock(&impl->sync);
+        aws_condition_variable_wait_pred(&impl->pending_line_signal, &impl->sync, s_background_wait_fn, impl);
+
+        size_t line_count = aws_array_list_length(&impl->pending_log_lines);
+        bool finished = impl->finished;
+
+        if (line_count == 0) {
+            aws_mutex_unlock(&impl->sync);
+            if (finished) {
+                break;
+            }
+            continue;
+        }
+
+        aws_array_list_swap_contents(&impl->pending_log_lines, &log_lines);
+        aws_mutex_unlock(&impl->sync);
+
+        /*
+         * Consider copying these into a page-sized stack buffer (string) and then making the write calls
+         * against it rather than the individual strings.  Might be a savings when > 1 lines (cut down on
+         * write calls).
+         */
+        for (size_t i = 0; i < line_count; ++i) {
+            struct aws_string *log_line = NULL;
+            if (aws_array_list_get_at(&log_lines, &log_line, i)) {
+                continue;
+            }
+
+            (*channel->writer->vtable->write_fn)(channel->writer, log_line);
+
+            /*
+             * send is considered a transfer of ownership.  write is not a transfer of ownership.
+             * So it's always the channel's responsibility to clean up all log lines that enter
+             * it as soon as they are no longer needed.
+             */
+            aws_string_destroy(log_line);
+        }
+
+        aws_array_list_clear(&log_lines);
+    }
+
+    aws_array_list_clean_up(&log_lines);
+}
+
+int aws_background_log_channel_init(struct aws_log_channel *channel, struct aws_allocator *allocator, struct aws_log_writer *writer) {
+    struct aws_log_background_channel *impl = (struct aws_log_background_channel *)aws_mem_acquire(allocator, sizeof(struct aws_log_background_channel));
+    if (impl == NULL) {
+        return aws_raise_error(AWS_ERROR_OOM);
+    }
+
+    impl->finished = false;
+
+    if (aws_mutex_init(&impl->sync)) {
+        goto cleanup_sync_init_fail;
+    }
+
+    if (aws_array_list_init_dynamic(&impl->pending_log_lines, allocator, 10, sizeof(struct aws_string *))) {
+        goto cleanup_pending_log_lines_init_fail;
+    }
+
+    if (aws_condition_variable_init(&impl->pending_line_signal)) {
+        goto cleanup_pending_line_signal_init_fail;
+    }
+
+    if (aws_thread_init(&impl->background_thread, allocator)) {
+        goto cleanup_background_thread_init_fail;
+    }
+
+    channel->vtable = &s_background_channel_vtable;
+    channel->allocator = allocator;
+    channel->impl = impl;
+    channel->writer = writer;
+
+    /*
+     * Logging thread should need very little stack, but let's defer this to later
```

Comment:
Add a SIM ticket to follow up on this, or we'll forget for sure :)

Reply:
Done